status/status.go

```diff
@@ -128,3 +130,39 @@ func FromError(err error) (s *Status, ok bool) {
 	}
 	return nil, false
 }
+
+// WithDetails returns a new status with the provided details messages appended to the status.
+// If any errors are encountered, it returns nil and the first error encountered.
+func (s *Status) WithDetails(details ...proto.Message) (*Status, error) {
+	if s.Code() == codes.OK {
+		return nil, errors.New("no error details for status with code OK")
+	}
+	// s.Code() != OK implies that s.Proto() != nil.
+	p := s.Proto()
+	for _, detail := range details {
+		any, err := ptypes.MarshalAny(detail)
+		if err != nil {
+			return nil, err
+		}
+		p.Details = append(p.Details, any)
+	}
+	return &Status{s: p}, nil
+}
+
+// Details returns a slice of details messages attached to the status.
+// If any errors are encountered, it returns nil and the first error encountered.
+func (s *Status) Details() ([]proto.Message, error) {
+	if s.Code() == codes.OK {
+		return nil, errors.New("no error details for status with code OK")
+	}
+	// s.Code() != OK implies that s.s != nil.
+	details := make([]proto.Message, 0, len(s.s.Details))
+	for _, any := range s.s.Details {
+		detail := &ptypes.DynamicAny{}
+		if err := ptypes.UnmarshalAny(any, detail); err != nil {
+			return nil, err
+		}
```

Comment:
I agree with Michael. This is an unusual case, because the dynamic nature of Any means there's a reasonable chance that an old client might not have the right protos linked in, and partial details are still likely to be useful.

While it's true that returning `nil` with errors is a best practice, it's not a hard requirement, and there are counterexamples. The most notable is `io.Writer.Write`, which returns the number of bytes written along with an error.